sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -737,6 +737,10 @@ abstract class PushableColumnBase {
         }
       case s: GetStructField if nestedPredicatePushdownEnabled =>
         helper(s.child).map(_ :+ s.childSchema(s.ordinal).name)
+      case GetJsonObject(col, field) =>
+        val colName = col.toString.split("#")(0)
+        val fieldName = field.toString.split("\\.").last
+        Some(Seq("GetJsonObject(" + col + "," + field + ")"))
```

Comment:
How do we distinguish the actual field name vs the pushed down `GetJsonObject` in the Data source implementation side?

Reply:
I gave an example for downstream implementation here:

https://issues.apache.org/jira/browse/SPARK-33915?focusedCommentId=17257647&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17257647

Reply:
I know it is rather rare case but how do we distinguish `GetJsonObject(col, field)` as a field name vs as the expression pushed down? I doubt if this is a general way that we can leverage to push expressions down.

Reply:
Can you outline an example where GetJsonObject(col, field) is the field name ?

thanks

Reply:
For example,

```
scala> spark.range(1).toDF("GetJsonObject(col, field)").show()
+-------------------------+
|GetJsonObject(col, field)|
+-------------------------+
|                        0|
+-------------------------+


scala> spark.range(1).selectExpr("id as `GetJsonObject(col, field)`").show()
+-------------------------+
|GetJsonObject(col, field)|
+-------------------------+
|                        0|
+-------------------------+
```

Reply:
Please note that the actual column and field names passed to GetJsonObject() look like the following:
```
scala> spark.range(1).toDF("GetJsonObject(phone#37,$.phone)").show()
java.lang.IllegalArgumentException: Error while instantiating 'org.apache.spark.sql.hive.HiveSessionStateBuilder':
  at org.apache.spark.sql.SparkSession$.org$apache$spark$sql$SparkSession$$instantiateSessionState(SparkSession.scala:1115)
  at org.apache.spark.sql.SparkSession.$anonfun$sessionState$2(SparkSession.scala:158)
  at scala.Option.getOrElse(Option.scala:189)
  at org.apache.spark.sql.SparkSession.sessionState$lzycompute(SparkSession.scala:156)
  at org.apache.spark.sql.SparkSession.sessionState(SparkSession.scala:153)
  at org.apache.spark.sql.Dataset.<init>(Dataset.scala:217)
  at org.apache.spark.sql.SparkSession.range(SparkSession.scala:538)
  at org.apache.spark.sql.SparkSession.range(SparkSession.scala:517)
  at org.apache.spark.sql.SparkSession.range(SparkSession.scala:508)
  ... 47 elided
Caused by: java.lang.IllegalArgumentException: wrong number of arguments
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  at org.apache.spark.sql.SparkSession$.org$apache$spark$sql$SparkSession$$instantiateSessionState(SparkSession.scala:1112)
  ... 55 more
```
I tried some variation of the above but couldn't get the statement to pass.

Reply:
How can we make sure all the data sources support `GetJsonObject` function?

Reply:
The above question is orthogonal to @HyukjinKwon 's question, let me resolve that question and answer using another comment.

Reply:
@tedyu, I think there was something wrong when you build Spark. Just for doubly sure, downloading the latest Spark and running it works as below:

```
scala> spark.range(1).toDF("GetJsonObject(phone#37,$.phone)").show()
+-------------------------------+
|GetJsonObject(phone#37,$.phone)|
+-------------------------------+
|                              0|
+-------------------------------+
```

Reply:
I rebuilt Spark locally and was able to get the same output for the query you gave above.

Please note that get_json_object should be used here instead of GetJsonObject().

I tried the following:
```
scala> spark.sql("CREATE TABLE IF NOT EXISTS t(id int , phone string)")
...
scala> spark.sql("select get_json_object(phone,'$.code') from t")
res10: org.apache.spark.sql.DataFrame = [get_json_object(phone, $.code): string]
```
If an invalid column is specified:
```
scala> spark.sql("select get_json_object(phon,'$.code') from t")
org.apache.spark.sql.AnalysisException: cannot resolve '`phon`' given input columns: [spark_catalog.default.t.id, spark_catalog.default.t.phone]; line 1 pos 23;
'Project [unresolvedalias('get_json_object('phon, $.code), None)]
+- SubqueryAlias spark_catalog.default.t
   +- HiveTableRelation [`default`.`t`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [id#17, phone#18], Partition Cols: []]
```

Reply:
@tedyu, it should be backquoted because it contains special characters:

```scala
spark.sql("select `GetJsonObject(phone,'$.code')` from t")
```

I am more asking how the downstream DSv2 implementation should handle the string `GetJsonObject(phone,'$.code')`. Is it a field name or pushed JSON expression? This seems ambiguous.

Reply:
Not sure if you ran the above statement in a shell. I got the following:
```
scala> spark.sql("select * from t")
res2: org.apache.spark.sql.DataFrame = [id: int, phone: string]

org.apache.spark.sql.AnalysisException: cannot resolve '`GetJsonObject(phone,'$.code')`' given input columns: [spark_catalog.default.t.id, spark_catalog.default.t.phone]; line 1 pos 7;
'Project ['`GetJsonObject(phone,'$.code')`]
+- SubqueryAlias spark_catalog.default.t
   +- HiveTableRelation [`default`.`t`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [id#0, phone#1], Partition Cols: []]

  at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.failAnalysis(package.scala:42)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$$nestedInanonfun$checkAnalysis$1$2.applyOrElse(CheckAnalysis.scala:170)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$$nestedInanonfun$checkAnalysis$1$2.applyOrElse(CheckAnalysis.scala:167)
  at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformUp$2(TreeNode.scala:341)
  at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:73)
  at org.apache.spark.sql.catalyst.trees.TreeNode.transformUp(TreeNode.scala:341)
  at org.apache.spark.sql.catalyst.plans.QueryPlan.$anonfun$transformExpressionsUp$1(QueryPlan.scala:104)
  at org.apache.spark.sql.catalyst.plans.QueryPlan.$anonfun$mapExpressions$1(QueryPlan.scala:116)
  at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:73)
  at org.apache.spark.sql.catalyst.plans.QueryPlan.transformExpression$1(QueryPlan.scala:116)
  at org.apache.spark.sql.catalyst.plans.QueryPlan.recursiveTransform$1(QueryPlan.scala:127)
  at org.apache.spark.sql.catalyst.plans.QueryPlan.$anonfun$mapExpressions$3(QueryPlan.scala:132)
  at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
  at scala.collection.immutable.List.foreach(List.scala:392)
  at scala.collection.TraversableLike.map(TraversableLike.scala:238)
  at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
  at scala.collection.immutable.List.map(List.scala:298)
  at org.apache.spark.sql.catalyst.plans.QueryPlan.recursiveTransform$1(QueryPlan.scala:132)
  at org.apache.spark.sql.catalyst.plans.QueryPlan.$anonfun$mapExpressions$4(QueryPlan.scala:137)
  at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:243)
  at org.apache.spark.sql.catalyst.plans.QueryPlan.mapExpressions(QueryPlan.scala:137)
  at org.apache.spark.sql.catalyst.plans.QueryPlan.transformExpressionsUp(QueryPlan.scala:104)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis$1(CheckAnalysis.scala:167)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis$1$adapted(CheckAnalysis.scala:94)
  at org.apache.spark.sql.catalyst.trees.TreeNode.foreachUp(TreeNode.scala:183)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.checkAnalysis(CheckAnalysis.scala:94)
```
I don't think there is ambiguity.

Reply:
```
scala> spark.sql("CREATE TABLE IF NOT EXISTS t1(`GetJsonObject(phone,'$.code')`)")
org.apache.spark.sql.catalyst.parser.ParseException:
mismatched input ')' expecting {'ADD', 'AFTER', 'ALL', 'ALTER', 'ANALYZE', 'AND', 'ANTI', 'ANY', 'ARCHIVE', 'ARRAY', 'AS', 'ASC', 'AT', 'AUTHORIZATION', 'BETWEEN', 'BOTH', 'BUCKET', 'BUCKETS', 'BY', 'CACHE', 'CASCADE', 'CASE', 'CAST', 'CHANGE', 'CHECK', 'CLEAR', 'CLUSTER', 'CLUSTERED', 'CODEGEN', 'COLLATE', 'COLLECTION', 'COLUMN', 'COLUMNS', 'COMMENT', 'COMMIT', 'COMPACT', 'COMPACTIONS', 'COMPUTE', 'CONCATENATE', 'CONSTRAINT', 'COST', 'CREATE', 'CROSS', 'CUBE', 'CURRENT', 'CURRENT_DATE', 'CURRENT_TIME', 'CURRENT_TIMESTAMP', 'CURRENT_USER', 'DATA', 'DATABASE', DATABASES, 'DBPROPERTIES', 'DEFINED', 'DELETE', 'DELIMITED', 'DESC', 'DESCRIBE', 'DFS', 'DIRECTORIES', 'DIRECTORY', 'DISTINCT', 'DISTRIBUTE', 'DIV', 'DROP', 'ELSE', 'END', 'ESCAPE', 'ESCAPED', 'EXCEPT', 'EXCHANGE', 'EXISTS', 'EXPLAIN', 'EXPORT', 'EXTENDED', 'EXTERNAL', 'EXTRACT', 'FALSE', 'FETCH', 'FIELDS', 'FILTER', 'FILEFORMAT', 'FIRST', 'FOLLOWING', 'FOR', 'FOREIGN', 'FORMAT', 'FORMATTED', 'FROM', 'FULL', 'FUNCTION', 'FUNCTIONS', 'GLOBAL', 'GRANT', 'GROUP', 'GROUPING', 'HAVING', 'IF', 'IGNORE', 'IMPORT', 'IN', 'INDEX', 'INDEXES', 'INNER', 'INPATH', 'INPUTFORMAT', 'INSERT', 'INTERSECT', 'INTERVAL', 'INTO', 'IS', 'ITEMS', 'JOIN', 'KEYS', 'LAST', 'LATERAL', 'LAZY', 'LEADING', 'LEFT', 'LIKE', 'LIMIT', 'LINES', 'LIST', 'LOAD', 'LOCAL', 'LOCATION', 'LOCK', 'LOCKS', 'LOGICAL', 'MACRO', 'MAP', 'MATCHED', 'MERGE', 'MSCK', 'NAMESPACE', 'NAMESPACES', 'NATURAL', 'NO', NOT, 'NULL', 'NULLS', 'OF', 'ON', 'ONLY', 'OPTION', 'OPTIONS', 'OR', 'ORDER', 'OUT', 'OUTER', 'OUTPUTFORMAT', 'OVER', 'OVERLAPS', 'OVERLAY', 'OVERWRITE', 'PARTITION', 'PARTITIONED', 'PARTITIONS', 'PERCENT', 'PIVOT', 'PLACING', 'POSITION', 'PRECEDING', 'PRIMARY', 'PRINCIPALS', 'PROPERTIES', 'PURGE', 'QUERY', 'RANGE', 'RECORDREADER', 'RECORDWRITER', 'RECOVER', 'REDUCE', 'REFERENCES', 'REFRESH', 'RENAME', 'REPAIR', 'REPLACE', 'RESET', 'RESPECT', 'RESTRICT', 'REVOKE', 'RIGHT', RLIKE, 'ROLE', 'ROLES', 'ROLLBACK', 'ROLLUP', 'ROW', 'ROWS', 'SCHEMA', 'SELECT', 'SEMI', 'SEPARATED', 'SERDE', 'SERDEPROPERTIES', 'SESSION_USER', 'SET', 'MINUS', 'SETS', 'SHOW', 'SKEWED', 'SOME', 'SORT', 'SORTED', 'START', 'STATISTICS', 'STORED', 'STRATIFY', 'STRUCT', 'SUBSTR', 'SUBSTRING', 'TABLE', 'TABLES', 'TABLESAMPLE', 'TBLPROPERTIES', TEMPORARY, 'TERMINATED', 'THEN', 'TIME', 'TO', 'TOUCH', 'TRAILING', 'TRANSACTION', 'TRANSACTIONS', 'TRANSFORM', 'TRIM', 'TRUE', 'TRUNCATE', 'TYPE', 'UNARCHIVE', 'UNBOUNDED', 'UNCACHE', 'UNION', 'UNIQUE', 'UNKNOWN', 'UNLOCK', 'UNSET', 'UPDATE', 'USE', 'USER', 'USING', 'VALUES', 'VIEW', 'VIEWS', 'WHEN', 'WHERE', 'WINDOW', 'WITH', 'ZONE', IDENTIFIER, BACKQUOTED_IDENTIFIER}(line 1, pos 61)

== SQL ==
CREATE TABLE IF NOT EXISTS t1(`GetJsonObject(phone,'$.code')`)
-------------------------------------------------------------^^^

  at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver.scala:255)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:124)
  at org.apache.spark.sql.execution.SparkSqlParser.parse(SparkSqlParser.scala:50)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parsePlan(ParseDriver.scala:75)
  at org.apache.spark.sql.SparkSession.$anonfun$sql$2(SparkSession.scala:613)
  at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
  at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:613)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:610)
```